# Table of Contents 
### Import libraries and dfs
### Step 1: Perform consistency checks
### Step 2: Share if any data looks off or need investigation
### Step 3: Check for mixed-type data
### Step 4: Fix any mixed-type data
### Step 5: Run check for missing values
### Step 6: Address missing values
### Step 7: Run check for duplicate values
### Step 8: Address duplicate values
### Step 9: Export dfs

### Import libraries and dfs

In [2]:
# import libraries 
import pandas as pd
import numpy as np
import os

In [3]:
# turning folder path into string
path = r'/Users/davesmac/Desktop/04-2022- Instacart Basket Analysis'

In [4]:
#import products via os library
df_prods = pd.read_csv(os.path.join(path, '02 Data', 'Original Data', 'products.csv'), index_col = False)

In [5]:
#import orders via os library
df_ords = pd.read_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = False)

In [6]:
#import ords_prods_merge
ords_prods_merge = pd.read_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))

In [12]:
#convert values in 'prices' >100 to 'nan'
ords_prods_merge.loc[ords_prods_merge['prices'] > 100] = np.nan

In [7]:
ords_prods_merge.isnull().sum()

order_id                     5127
user_id                      5127
order_number                 5127
orders_day_of_week           5127
order_hour_of_day            5127
days_since_prior_order    2080872
first_order                  5127
product_id                   5127
add_to_cart_order            5127
reordered                    5127
product_name                 5127
aisle_id                     5127
department_id                5127
prices                       5127
_merge                       5127
price_label                  5127
Busiest_days                 5127
busiest_period_of_day        5127
max_order                    5127
loyalty_flag                 5127
avg_price_per_order          5127
spender_flag                 5127
order_frequency              5132
order_frequency_flag         5132
dtype: int64

# Consistency Checks on df_prods

## Step 1: Perform consistency checks

In [6]:
# Finding missing values in df_prods
df_prods.isnull().sum()

product_id        0
product_name     16
aisle_id          0
department_id     0
prices            0
dtype: int64

In [7]:
# Create subset showing only rows with missing values in product_name
df_nan = df_prods[df_prods['product_name'].isnull() == True]

In [8]:
df_nan

product_id product_name  aisle_id  department_id  prices
33             34          NaN       121             14    12.2
68             69          NaN        26              7    11.8
115           116          NaN        93              3    10.8
261           262          NaN       110             13    12.1
525           525          NaN       109             11     1.2
1511         1511          NaN        84             16    14.3
1780         1780          NaN       126             11    12.3
2240         2240          NaN        52              1    14.2
2586         2586          NaN       104             13    12.4
3159         3159          NaN       126             11    13.1
3230         3230          NaN       120             16    14.4
3736         3736          NaN        41              8    14.8
4283         4283          NaN        77              7    14.4
4790         4790          NaN        91             16    14.5
38187       38183          NaN        39             12    20.9
40444       40440          NaN       120             16    14.8

In [9]:
# Number of rows in df_prods
df_prods.shape

(49693, 5)

In [11]:
# Create subset w/out missing values in product_name
df_prods_clean = df_prods[df_prods['product_name'].isnull() == False]

In [12]:
# Number of rows after omitting missing values
df_prods_clean.shape

(49677, 5)

In [13]:
# Create subset showing only duplicate values
df_dups = df_prods_clean[df_prods_clean.duplicated()]

In [14]:
df_dups

product_id                                       product_name  \
462           462                  Fiber 4g Gummy Dietary Supplement   
18459       18458                                         Ranger IPA   
26810       26808               Black House Coffee Roasty Stout Beer   
35309       35306  Gluten Free Organic Peanut Butter & Chocolate ...   
35495       35491                            Adore Forever Body Wash   

       aisle_id  department_id  prices  
462          70             11     4.8  
18459        27              5     9.2  
26810        27              5    13.4  
35309       121             14     6.8  
35495       127             11     9.9

In [15]:
# Create subset with dropped duplicates
df_prods_clean_no_dups = df_prods_clean.drop_duplicates()

In [16]:
# Number of rows in df_prods after removing missing values and dropping duplicates
df_prods_clean_no_dups.shape

(49672, 5)

## Step 2: Share if any data looks off or need investigation

In [17]:
#Descriptive stats on clean df_prods
df_prods_clean_no_dups.describe()

product_id      aisle_id  department_id        prices
count  49672.000000  49672.000000   49672.000000  49672.000000
mean   24850.349775     67.762442      11.728942      9.993282
std    14340.705287     38.315784       5.850779    453.615536
min        1.000000      1.000000       1.000000      1.000000
25%    12432.750000     35.000000       7.000000      4.100000
50%    24850.500000     69.000000      13.000000      7.100000
75%    37268.250000    100.000000      17.000000     11.100000
max    49688.000000    134.000000      21.000000  99999.000000

### Investigate further- min/max values of prices. The min value is 1, while the max is 99999. While an item could be 1 dollar, it seems very unlikely that there would be an item priced at 99999 dollars. 

# Consistency Checks on df_ords

## Step 3: Check for mixed-data type

In [18]:
# Check for mixed types in df_ords
for col in df_ords.columns.tolist():
  weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
  if len (df_ords[weird]) > 0:
    print (col)

## Step 4: Fix mixed-type data

### No mixed-type data found

## Step 5: Run check for missing values


In [19]:
# Finding missing values in df_ords
df_ords.isnull().sum()

Unnamed: 0                     0
order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

### Missing values have been found in the days_since_prior_order column. This is most likley due to to the fact that every customer at one point placed their first order, which obviously did not have an order placed before it. This results in missing values in said column.

## Step 6: Address missing values

In [32]:
# Looking at rows with missing data in days_since_prior_order column
# Create a new subset with this data
df_ords_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [42]:
# New subset shows rows with missing data are all order number 1. 
df_ords_nan.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0            0   2539329        1             1                   2   
11          11   2168274        2             1                   2   
26          26   1374495        3             1                   1   
39          39   3343014        4             1                   6   
45          45   2717275        5             1                   3   

    order_hour_of_day  days_since_prior_order  first_order  
0                   8                     NaN         True  
11                 11                     NaN         True  
26                 14                     NaN         True  
39                 11                     NaN         True  
45                 12                     NaN         True

In [43]:
df_ords_nan.tail()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
3420930     3420930    969311   206205             1                   4   
3420934     3420934   3189322   206206             1                   3   
3421002     3421002   2166133   206207             1                   6   
3421019     3421019   2227043   206208             1                   1   
3421069     3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  first_order  
3420930                 12                     NaN         True  
3420934                 18                     NaN         True  
3421002                 19                     NaN         True  
3421019                 15                     NaN         True  
3421069                 11                     NaN         True

In [22]:
# Create new dataframe for clean data.
df_ords_clean = df_ords

In [23]:
# Create new column for customer's first order
df_ords_clean['first_order'] = df_ords['days_since_prior_order'].isnull() == True

In [24]:
df_ords_clean.head()

Unnamed: 0  order_id  user_id  order_number  orders_day_of_week  \
0           0   2539329        1             1                   2   
1           1   2398795        1             2                   3   
2           2    473747        1             3                   3   
3           3   2254736        1             4                   4   
4           4    431534        1             5                   4   

   order_hour_of_day  days_since_prior_order  first_order  
0                  8                     NaN         True  
1                  7                    15.0        False  
2                 12                    21.0        False  
3                  7                    29.0        False  
4                 15                    28.0        False

### I decided to create a new column stating whether or not this was a customers first order. I did this because I found that the rows with missing values in days_since_prior_order were all the first order for any given customer. When the value in the new first_order comes back as 'True', we know that that is a customers first order and thus why there is a missing value. 

### I decided not to just replace missing values with '0' or the mean/median value since that could provide misleading info. 

## Step 7: Run check for duplicate values

In [25]:
# Checking for duplicate records
df_dups = df_ords_clean[df_ords_clean.duplicated()]

In [26]:
df_dups

Empty DataFrame
Columns: [Unnamed: 0, order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order, first_order]
Index: []

### No duplicate records have been found for df_orders_clean

## Step 8: Address duplicate values

### No duplicate records have been found- do not need to address

## Step 9: Export dfs

In [38]:
# Export df_prods_clean_no_dups to prepared data
df_prods_clean_no_dups.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'products_checked.csv'))

In [40]:
# Export df_ords_clean to prepared data
df_ords_clean.to_csv(os.path.join(path, '02 Data', 'Prepared Data', 'orders_checked.csv'))

In [13]:
#exporting df as a pkl file
ords_prods_merge.to_pickle(os.path.join(path, '02 Data', 'Prepared Data', 'orders_products_merged.pkl'))